src/main/java/ladysnake/effective/client/world/WaterfallCloudGenerators.java

```diff
@@ -64,13 +64,14 @@ public static void tick() {
     }
 
     private static void tickParticles(World world) {
-        for (BlockPos pos : particlesToSpawn.keySet()) {
-            if (pos != null) {
-                if (particlesToSpawn.put(pos, particlesToSpawn.getInt(pos) - 1) <= 0) {
-                    particlesToSpawn.removeInt(pos);
+        synchronized (particlesToSpawn) {
+            for (BlockPos pos : particlesToSpawn.keySet()) {
+                if (pos != null) {
+                    particlesToSpawn.put(pos, particlesToSpawn.getInt(pos) - 1);
```

Comment:
The `computeX` family of methods exists to encapsulate the `put(x, get(x))` pattern.